### PR TITLE
Improve HTML form reset and submit

### LINKED ISSUE #1690: ComboBox in form submits input value rather than selected key
# 🐛 Bug Report 

Context: https://github.com/adobe/react-spectrum/discussions/1672

## 🤔 Expected Behavior

Not 100% sure but this is what I think might be correct:
ComboBox with `allowCustomValue=false` should probably submit its selected key on form submit. ComboBox with `allowCustomValue=true` should probably submit its input value on form submit if no selected key exists, otherwise submit its selected key.

EDIT: The above proposal has an issue since the user wouldn't be able to tell if it is a key or the input value being submitted

## 😯 Current Behavior

ComboBox always submits its input value on form submit

## 💁 Possible Solution

Possibly use HiddenSelect? Use something similar to HiddenSelect? Need to investigate

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| react-spectrum   |
| Browser          |
| Operating System |

### LINKED ISSUE #2825: Buttons with `type="reset"` don't reset the Form properly
# 🐛 Bug Report 

Some issues regarding Form reseting via buttons with type="reset":
- Enter/Space don't trigger the form reset
- Controlled components in React don't play well with native reset. Resolve this by adding `onReset` so user can reset their controlled components properly
- Check components that are controlled under the hood by default (e.g. NumberField)

Add docs when the above are fixed

## 🤔 Expected Behavior

Forms should be reset if a button w/ type="reset" is pressed

## 😯 Current Behavior

Forms work with buttons of type="reset". `onReset` should be a supported callback

## 💁 Possible Solution

PR in progress: https://github.com/adobe/react-spectrum/pull/2824

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| react-spectrum   |
| Browser          |
| Operating System |

## 🧢 Your Company/Team

RSP

### PR DESCRIPTION
**Pull request by devongovett**
Closes #1690, closes #2825

This improves support for HTML form reset and submit across many components that didn't previously support it. Extracted from #4640. Form validation support is being reworked a bit and will come in a separate PR.

* Adds support for native form reset, e.g. `<Button type="reset">`. This is implemented by subscribing to the `reset` event on the input's parent form, and triggering an onChange with the initially rendered value. Works with both uncontrolled and controlled inputs.
* Adds support for submitting the numeric value in NumberField rather than the formatted text. Previously the `name` prop wasn't passed through at all, so this is not a breaking change. It is implemented by rendering an extra `<input type="hidden">`.
* Adds support for `name` prop to slider and color components.
* Adds support for submitting a stringified version of the date value in all date picker components, using an `<input type="hidden">`.
* Adds support for submitting the key rather than the text value of a ComboBox, also using an `<input type="hidden">`. This can be controlled by setting the `formValue` prop. When `allowsCustomValue` is true, we always submit the text value.

### CODE REVIEW

File: packages/react-aria-components/src/ComboBox.tsx [hunk 1]
```diff
@@ -49,16 +49,30 @@ export interface ComboBoxRenderProps {
...
 }
 
 export const ComboBoxContext = createContext<ContextValue<ComboBoxProps<any>, HTMLDivElement>>(null);
 
 function ComboBox<T extends object>(props: ComboBoxProps<T>, ref: ForwardedRef<HTMLDivElement>) {
   [props, ref] = useContextProps(props, ref, ComboBoxContext);
   let [propsFromListBox, setListBoxProps] = useState<ListBoxProps<T>>({children: []});
+  let {
+    name,
+    formValue = 'key',
```
Comment (devongovett) [inline packages/react-aria-components/src/ComboBox.tsx:70]: Note: this is intentionally different from RSP. I think submitting the key will probably be more common, but doing so in RSP would be a breaking change. Thoughts?

Comment (reidbarber) [reply packages/react-aria-components/src/ComboBox.tsx:70]: Agreed that `key` seems like a better default. That sounds reasonable to me.

**Review by reidbarber — APPROVED** [review]
LGTM

File: packages/react-aria-components/src/ComboBox.tsx (hunk 1 above)
Comment (ktabors) [reply packages/react-aria-components/src/ComboBox.tsx:70]: I agree that submitting a key would be more common and more useful.

**Review by ktabors — APPROVED** [review]

**Pull request merged**